plugins/inputs/snmp2/snmp.go

```diff
@@ -0,0 +1,788 @@
+	}
+	for _, r := range rows {
+		if len(r.Tags) < tagCount {
+			// don't add rows which are missing tags, as without tags you can't filter
+			continue
+		}
+		rt.Rows = append(rt.Rows, r)
+	}
+	return &rt, nil
+}
+
+// snmpConnection is an interface which wraps a *gosnmp.GoSNMP object.
+// We interact through an interface so we can mock it out in tests.
+type snmpConnection interface {
+	Host() string
+	//BulkWalkAll(string) ([]gosnmp.SnmpPDU, error)
+	Walk(string, gosnmp.WalkFunc) error
+	Get(oids []string) (*gosnmp.SnmpPacket, error)
+}
+
+// gosnmpWrapper wraps a *gosnmp.GoSNMP object so we can use it as a snmpConnection.
+type gosnmpWrapper struct {
+	*gosnmp.GoSNMP
+}
+
+// Host returns the value of GoSNMP.Target.
+func (gsw gosnmpWrapper) Host() string {
+	return gsw.Target
+}
+
+// Walk wraps GoSNMP.Walk() or GoSNMP.BulkWalk(), depending on whether the
+// connection is using SNMPv1 or newer.
+// Also, if any error is encountered, it will just once reconnect and try again.
+func (gsw gosnmpWrapper) Walk(oid string, fn gosnmp.WalkFunc) error {
+	var err error
+	// On error, retry once.
+	// Unfortunately we can't distinguish between an error returned by gosnmp, and one returned by the walk function.
+	for i := 0; i < 2; i++ {
+		if gsw.Version == gosnmp.Version1 {
+			err = gsw.GoSNMP.Walk(oid, fn)
+		} else {
+			err = gsw.GoSNMP.BulkWalk(oid, fn)
+		}
+		if err == nil {
+			return nil
+		}
+		if err := gsw.GoSNMP.Connect(); err != nil {
+			return Errorf(err, "reconnecting")
+		}
+	}
+	return err
+}
+
+// Get wraps GoSNMP.GET().
+// If any error is encountered, it will just once reconnect and try again.
+func (gsw gosnmpWrapper) Get(oids []string) (*gosnmp.SnmpPacket, error) {
+	var err error
+	var pkt *gosnmp.SnmpPacket
+	for i := 0; i < 2; i++ {
+		pkt, err = gsw.GoSNMP.Get(oids)
+		if err == nil {
+			return pkt, nil
+		}
+		if err := gsw.GoSNMP.Connect(); err != nil {
+			return nil, Errorf(err, "reconnecting")
+		}
+	}
+	return nil, err
+}
+
+// getConnection creates a snmpConnection (*gosnmp.GoSNMP) object and caches the
+// result using `agent` as the cache key.
+func (s *Snmp2) getConnection(agent string) (snmpConnection, error) {
+	if s.connectionCache == nil {
+		s.connectionCache = map[string]snmpConnection{}
+	}
+	if gs, ok := s.connectionCache[agent]; ok {
+		return gs, nil
+	}
+
+	gs := gosnmpWrapper{&gosnmp.GoSNMP{}}
+
+	host, portStr, err := net.SplitHostPort(agent)
+	if err != nil {
+		if err, ok := err.(*net.AddrError); !ok || err.Err != "missing port in address" {
+			return nil, Errorf(err, "parsing host")
+		}
+		host = agent
+		portStr = "161"
+	}
+	gs.Target = host
+
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, Errorf(err, "parsing port")
+	}
+	gs.Port = uint16(port)
+
+	if s.Timeout != "" {
+		if gs.Timeout, err = time.ParseDuration(s.Timeout); err != nil {
+			return nil, Errorf(err, "parsing timeout")
+		}
+	} else {
+		gs.Timeout = time.Second * 1
```

Comment:
This default seems to be a bit on the short side to me, especially when we're talking about SNMP and potentially very old and/or slow devices. Maybe default to 5s?